osu.Game/Graphics/UserInterfaceV2/DirectorySelector.cs

```diff
@@ -77,39 +77,75 @@ private void load()
             CurrentPath.BindValueChanged(updateDisplay, true);
         }
 
+        /// <summary>
+        /// Because <see cref="CurrentPath"/> changes may not necessarily lead to directories that exist/are accessible,
+        /// <see cref="updateDisplay"/> may need to change <see cref="CurrentPath"/> again to lead to a directory that is actually accessible.
+        /// This flag intends to prevent recursive <see cref="updateDisplay"/> calls from taking place during the process of finding an accessible directory.
+        /// </summary>
+        private bool directoryChanging;
+
         private void updateDisplay(ValueChangedEvent<DirectoryInfo> directory)
         {
+            if (directoryChanging)
+                return;
+
+            directoryChanging = true;
+
             directoryFlow.Clear();
 
-            try
+            var newDirectory = directory.NewValue;
+            bool flashError = false;
+            ICollection<DisplayPiece> displayPieces = Array.Empty<DisplayPiece>();
+
+            while (newDirectory != null)
             {
-                if (directory.NewValue == null)
-                {
-                    var drives = DriveInfo.GetDrives();
+                newDirectory.Refresh();
 
-                    foreach (var drive in drives)
-                        directoryFlow.Add(new DirectoryPiece(drive.RootDirectory));
-                }
-                else
-                {
-                    directoryFlow.Add(new ParentDirectoryPiece(CurrentPath.Value.Parent));
+                if (TryGetEntriesForPath(newDirectory, out displayPieces))
+                    break;
 
-                    directoryFlow.AddRange(GetEntriesForPath(CurrentPath.Value));
-                }
+                flashError = true;
+                newDirectory = newDirectory.Parent;
             }
-            catch (Exception)
-            {
-                CurrentPath.Value = directory.OldValue;
+
+            if (flashError)
                 this.FlashColour(Color4.Red, 300);
+
+            if (newDirectory == null)
+            {
+                var drives = DriveInfo.GetDrives();
+
+                foreach (var drive in drives)
+                    directoryFlow.Add(new DirectoryPiece(drive.RootDirectory));
+
+                return;
```

Comment:
Is `directoryChanging` not going to get stuck in an incorrect state here? Wondering if we want to use `try-finally` with the `directoryChanging = false;` inside the `finally` block here.